Including support for basic map icons

Taezen opened the pull request:
I have wrote a map manager that manages the map icons for the big map.

It is supporting the following map icons for now:
  - Quest icons (only pre-accepted ones, because accepted ones are getting handled by the client)
  - Guildleve icons
  - Guildhest icons
  - Beginner arena icons
  - Triple Triad Master npc icon

For the quest icons I reversed engineered the quest availability check from the client. It doesn't check everything what the client does for now, but all that what is currently supported by the server.

NotAdam commented on src/world/Manager/MapMgr.cpp:
```diff
@@ -0,0 +1,889 @@
+
+  if( quest.classJobUnlock )
+  {
+    if( quest.column18 == 3 )
```
this should be named instead of using the unknown value

NotAdam commented on src/world/Manager/MapMgr.cpp:
```diff
@@ -0,0 +1,889 @@
+      return false;
+  }
+
+  if( eventData.first == 65968 ) // Quest: A Legend for a Legend
```
things like this and basically everything after it should be deferred to a script, not inline

Taezen replied:
like an availability check in a quest script?

NotAdam replied:
yeah, probably something like `bool showMapIcon(Entity::Player&)`

NotAdam commented on src/world/Manager/MapMgr.cpp:
```diff
@@ -0,0 +1,889 @@
+void Sapphire::World::Manager::MapMgr::updateAll( Entity::Player& player )
+{
+  auto& mapData = m_mapData[ player.getTerritoryTypeId() ];
+  std::multimap< uint32_t, EventData, less > newMapData;
```
is there a particular reason for using a multimap instead of just a list with a defined structure?

Taezen replied:
The retail server seems to sort it by the lower word of the event id.
So I used a multimap for auto sorting with a custom sort method.
For lower processing time we can use a list here, because it seems the sorting is irrelevant for the client.

NotAdam replied:
yeah my main concern was that this obscures the behavior and makes it a bit more complicated for no real gain - if ordering was necessary though i would just opt to store them with the event id as they'll be implicitly ordered by that anyway if they're dumped into a set/map when the key is the event id

NotAdam commented on src/world/Manager/MapMgr.h:
```diff
@@ -0,0 +1,132 @@
+      uint32_t actorId;
+    };
+
+    struct QuestData
```
this should be using the struct in exddatagenerated

NotAdam reviewed (changes requested):
meant to get to this last week but have been busy, sorry

Taezen commented on src/world/Manager/MapMgr.cpp:
```diff
@@ -0,0 +1,889 @@
+              case EventHandler::EventHandlerType::CustomTalk:
+              {
+                // Include only the beginner arena icon yet. There a few other ones, that aren't referenced in the game files (Some examples are: The Triple Triad Tournament npc which has multiple icons and the ocean fishing icon)
+                if( npcData == 721223 )
```
Sorry, but I do not understand where the problem here is.
The custom talk icons have a very special handling from the server and not all of them are referenced in the exd data (The server decides which icon is displayed on the map. This applies to all supported events). So I have included this icon for now only.

NotAdam replied:
i think i started writing something a few days ago and then didn't finish, i think what i intended to write here is that this shouldn't be hardcoded in some random file but i can't be sure lol

Taezen replied:
Aah, you're right, that this should'nt be hardcoded, but I think this will do it for now, until it is figured out which custom talk icons are shown by retail. In addition it is unclear for me now which icons are displayed for custom talk npcs that aren't referenced in the exd data.

NotAdam commented on src/world/Manager/MapMgr.cpp:
```diff
@@ -0,0 +1,889 @@
+      }
+    }
+  }
+  std::cout << "\n";
```
also use the logger to write out instead of iostream

Taezen replied:
Ok I change that, but I thought this is ok, because in the `InstanceObjectCache` initialization it's done in the same way.

NotAdam replied:
more just some debugging code that was left behind lol

Taezen commented on src/common/Network/PacketDef/Zone/ServerZoneDef.h:
```diff
@@ -928,12 +928,12 @@ namespace Sapphire::Network::Packets::Server
     uint32_t unknown11;
-    uint32_t unknown12[4];
+    uint32_t unknown12[3];
+    uint32_t additionalFestivalId;
```
This is an InitZone packet from the Gold Saucer event:
`10 00 90 00 00 00 00 00 00 00 00 00 17 46 56 00`
`02 59 00 00 AA D8 CF 00 02 00 00 00 00 80 09 41`
`00 00 80 3F` **73 00 00 00** `00 00 00 00 00 00 00 00`
`00 00 00 00` **73 00 00 00** `00 00 00 00 00 00 00 00`
`00 00 00 00 59 1D 0D C2 F9 D5 23 3F B7 FF C7 42`
`00 01 01 04 04 01 04 03 01 04 02 00 00 00 00 00`

This is an InitZone packet from the Moonfaire event:
`17 00 89 00 00 00 00 00 00 00 00 00 C0 1E 81 00`
`02 39 10 00 AA E0 24 00 15 00 00 00 00 80 09 41`
`00 00 80 3F` **74 00 00 00** `00 00 00 00 00 00 00 00`
`00 00 00 00 `**74 00 00 00** `00 00 00 00 00 00 00 00`
`00 00 00 00 7B 94 75 43 2B DA 97 42 93 85 4D 44`
`00 01 04 04 01 01 01 04 01 02 01 00 00 00 00 00`

Because of these two festival InitZone packets, I think they have changed something in the festival system.

Taezen commented on src/world/Manager/MapMgr.cpp:
```diff
@@ -0,0 +1,889 @@
+    if( exdData.getField< std::string >( quest.second, 1 ).empty() ) // id
+      continue;
+
+    auto& questData = m_questData[ quest.first ];
```
I've tried it in this way first, but it takes approximately 10-20 ms to load the quest data directly from the exd. This is way too long, if you want to do it after the ZoneIn trigger like the retail server does.
Even if I remove the unused vectors from the Quest struct, it isn't really faster.
There are zones which contains over 100 quests to check. That's over a second processing time.
I didn't found any other way to this faster without storing them in memory.

NotAdam replied:
ah yeah, okay, i'm not surprised, datreader needs more than a bit of love in some areas

you can keep them in memory though, its just loading everything upfront is a bit costly when working on the server - i don't mind it that much as when i tried it, it wasn't too bad but its nice to avoid doing everything upfront where reasonable

Taezen commented on src/world/Manager/MapMgr.cpp:
```diff
@@ -0,0 +1,889 @@
+
+  std::set< std::string > bgSet;
+
+  for( auto territoryTypeId : exdData.getTerritoryTypeIdList() )
```
I didn't see any event nor npc data in the InstanceObjectCache or did I miss something?

NotAdam replied:
ah, they're filtered out - you can add them in here https://github.com/SapphireServer/Sapphire/blob/develop/src/world/Territory/InstanceObjectCache.cpp#L80

Taezen commented:
I have changed it to actually use the struct from `ExdDataGenerated.h` for the quest data.
It now caches only the quest data during initialization. All other is generated during runtime or fetched from the `InstanceObjectCache`.
I also removed the hardcoded quest checks and implemented a basic method for an availability check.
Unfortunately I had to hardcode the data for `GuildLeveAssignment`, because exd data isn't accessible from scripts right now.

NotAdam commented on src/world/Event/EventHandler.h:
```diff
@@ -85,6 +85,13 @@ namespace Sapphire::Event
       QuestBattleDirector = 0x8006,
     };
 
+    enum class EventState : uint8_t
```
i feel like `EventState` is a bit ambiguous, perhaps `EventAvailability`?

NotAdam commented on src/world/Script/NativeScriptApi.cpp:
```diff
@@ -138,6 +137,11 @@ namespace Sapphire::ScriptAPI
   {
   }
 
+  Event::EventHandler::EventState EventScript::checkExtraConditions( Sapphire::Entity::Player& player, uint32_t eventId )
```
probably want this to replicate the name of the struct a bit more, `checkExtraConditions` doesn't really mean anything obvious

Taezen replied:
I think this should be a function that checks requirements that aren't in the exd data. So this should called after the generic availability check.
So are something like `checkSpecialRequirements` are a better name for this?
Function names aren't my strength xD

NotAdam replied:
i thought about this a bit and i would actually prefer that this is just called `getQuestAvailability` and the return is just like `QuestAvailability` - i can't think of any cases where this would apply to anything that isn't a quest in this case, and we can call different handlers for different cases as needed

i'd rather have more specific handlers that have a very specific scope rather than one general function that does a bunch of different things because that'll just be a nightmare over time

Taezen replied:
Should this enum then stay in `EventHandler` or should I rather move it to `Common`?

NotAdam replied:
nah its fine where it is, it doesn't really make sense for it to be in common anyway given its only really used within the event system

NotAdam reviewed (approved):
looks good - nice work

Taezen commented:
Thanks :)
